File: Java/Example 3 (Advanced) - Export Trade History from MT4 to CSV.java
// Example 3 (Advanced) - Export Trade History from MT4 to CSV
//
// PREREQUISITE: Install MTsocketAPI in your Metatrader 4 using the following link https://www.mtsocketapi.com/doc4/installation.html
//
// WARNING: All these source codes are only examples for testing purposes. 
// WARNING: We don’t provide any guarantee or responsibility about it. 
// WARNING: Use these examples at your own risk.

import java.net.*;
import java.io.*;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;
/**
 *  $ javac -cp .:json-simple-1.1.1.jar getHistory.java (Compile)
 *  $ java -cp .:json-simple-1.1.1.jar getHistory (Run)
 * */
class getHistory
{
    public static void main(String[] args)
    {
           try
           {
                Socket socket = new Socket("localhost", 77);

                OutputStream output = socket.getOutputStream();
                InputStream input = socket.getInputStream();

                BufferedReader reader = new BufferedReader(new InputStreamReader(input));
                BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(output));

                String jsonString;
                JSONObject jo = new JSONObject();
                jo.put("MSG","TRADE_HISTORY");
                jo.put("FROM_DATE","2022/04/13 07:00:00");
                jo.put("TO_DATE","2022/04/15 00:00:00");

                writer.write(jo.toString() + "\r\n");
                writer.flush();

                jsonString = reader.readLine();
                System.out.println(jsonString);

                JSONParser parser = new JSONParser();
                JSONObject json = (JSONObject) parser.parse(jsonString);
                JSONArray jarr= (JSONArray)json.get("TRADES");

                FileWriter myFile = new FileWriter("tradeHistory.csv");

                myFile.write("OPEN_TIME,CLOSE_TIME,TICKET,SYMBOL,MAGIC,OPEN_PRICE,CLOSE_PRICE,TYPE,LOTS,STOP_LOSS,TAKE_PROFIT,SWAP,COMMISSION,COMMENT,PROFIT\r\n");

                for(int i = 0; i < jarr.size(); i++) {
                        JSONObject item = (JSONObject) jarr.get(i);

                        myFile.write(item.get("OPEN_TIME").toString() + "," +
                                item.get("CLOSE_TIME").toString() + "," +
                                item.get("TICKET").toString() + "," +
                                item.get("SYMBOL").toString() + "," +
                                item.get("MAGIC").toString() + "," +
                                item.get("OPEN_PRICE").toString() + "," +
                                item.get("CLOSE_PRICE").toString() + "," +
                                item.get("TYPE").toString() + "," +
                                item.get("LOTS").toString() + "," +
                                item.get("STOP_LOSS").toString() + "," +
                                item.get("TAKE_PROFIT").toString() + "," +
                                item.get("SWAP").toString() + "," +
                                item.get("COMMENT") + "," +
                                item.get("PROFIT").toString() +
                                "\r\n");
                }

                myFile.close();

                System.out.println("tradeHistory.csv created!");

           }
           catch(ParseException pe) {
               System.out.println("position: " + pe.getPosition());
               System.out.println(pe);
           }
           catch (UnknownHostException ex) {
                System.out.println("Server not found: " + ex.getMessage());
           }
           catch (IOException ex) {
                System.out.println("I/O error: " + ex.getMessage());
           }
        }
}

File: NodeJS/Example 3 (Advanced) - Export Trade History from MT4 to CSV.js
// Example 3 (Advanced) - Export Trade History from MT4 to CSV
//
// PREREQUISITE: Install MTsocketAPI in your Metatrader 4 using the following link https://www.mtsocketapi.com/doc4/installation.html
//
// WARNING: All these source codes are only examples for testing purposes. 
// WARNING: We don’t provide any guarantee or responsibility about it. 
// WARNING: Use these examples at your own risk.

const Net = require('net');
const fs = require('fs');
const delimiter = ',';
const exportedFile = 'tradeHistory.csv';
const fromDate = "2022/04/12 13:00:00";
const toDate = "2022/04/15 00:00:00";
const client_cmd = new Net.Socket();

client_cmd.connect(77, "localhost", function() {
        const JSONobj = {};
        JSONobj["MSG"]="TRADE_HISTORY";
        JSONobj["FROM_DATE"]=fromDate;
        JSONobj["TO_DATE"]=toDate;
        client_cmd.write(JSON.stringify(JSONobj) + '\r\n');
});

client_cmd.on('data', function(chunk) {
        const JSONresult = JSON.parse(chunk.toString());
        const numTrades = Object.keys(JSONresult['TRADES']).length;
        console.log(`Number of trades: ${numTrades}`);
        
        //First we must write the CSV HEADER
        fs.writeFileSync('tradeHistory.csv','OPEN_TIME,CLOSE_TIME,TICKET,SYMBOL,LOTS,OPEN_PRICE,CLOSE_PRICE,PROFIT\r\n',{ flag: 'a+' });
        
        //Loop JSON array
        for(let i = 0; i < numTrades; i++) {
                //You can add more files like COMMENT, SWAP, COMMISSION...
                const OPEN_TIME = JSONresult['TRADES'][i]['OPEN_TIME'];
                const CLOSE_TIME = JSONresult['TRADES'][i]['CLOSE_TIME'];
                const TICKET = JSONresult['TRADES'][i]['TICKET'];
                const SYMBOL = JSONresult['TRADES'][i]['SYMBOL'];
                const LOTS = JSONresult['TRADES'][i]['LOTS'];
                const OPEN_PRICE = JSONresult['TRADES'][i]['OPEN_PRICE'];
                const CLOSE_PRICE = JSONresult['TRADES'][i]['CLOSE_PRICE'];
                const PROFIT = JSONresult['TRADES'][i]['PROFIT'];
                
                fs.writeFileSync(exportedFile, OPEN_TIME + delimiter + CLOSE_TIME + delimiter + TICKET + delimiter + SYMBOL + delimiter + LOTS + delimiter + OPEN_PRICE + delimiter + CLOSE_PRICE + delimiter + PROFIT + '\r\n',{ flag: 'a+' });
        }
        
        console.log(`Created the file ${exportedFile}!`);
        
        client_cmd.end();
});
